source/extensions/filters/network/zookeeper_proxy/decoder.cc

```diff
@@ -170,11 +176,23 @@ void DecoderImpl::decodeOnWrite(Buffer::Instance& data, uint64_t& offset) {
   const auto xid = helper_.peekInt32(data, offset);
   const auto xid_code = static_cast<XidCodes>(xid);
 
+  // Find the corresponding request for this XID.
+  const auto it = requests_by_xid_.find(xid);
+
+  std::chrono::milliseconds latency;
+
+  // If this fails, it's a server-side bug.
+  if (xid_code != XidCodes::WATCH_XID) {
+    ASSERT(it != requests_by_xid_.end());
```

Comment:
Should the comment go right above the `ASSERT` to make it more clear (looking at where it was copied over from).

Also, probably more of discussion for a follow-up PR but should we be more lenient of server-side errors and just emit an error metric/log instead of relying on `ASSERT`s which get stripped out in prod builds?

Reply:
given that it's due to bad responses from a server (not program invariant which is what i thought initially when i suggested it) im inclined to agree, but yeah it can probably be done in another PR